subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/serp_promo/SerpPromo.kt

```diff
@@ -0,0 +1,121 @@
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.anvil.annotations.ContributesMultibinding
+import com.squareup.anvil.annotations.ContributesTo
+import dagger.Lazy
+import dagger.Module
+import dagger.Provides
+import javax.inject.Inject
+import javax.inject.Qualifier
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import logcat.logcat
+
+interface SerpPromo {
+    suspend fun injectCookie(cookieValue: String?)
+}
+
+private const val HTTPS_WWW_SUBSCRIPTION_DDG_COM = ".subscriptions.duckduckgo.com"
+private const val SERP_PPRO_PROMO_COOKIE_NAME = "privacy_pro_access_token"
+
+@ContributesBinding(
+    scope = AppScope::class,
+    boundType = SerpPromo::class,
+)
+@ContributesMultibinding(
+    scope = AppScope::class,
+    boundType = MainProcessLifecycleObserver::class,
+)
+class RealSerpPromo @Inject constructor(
+    @InternalApi private val cookieManager: CookieManagerWrapper,
+    private val dispatcherProvider: DispatcherProvider,
+    private val privacyProFeature: Lazy<PrivacyProFeature>,
+) : SerpPromo, MainProcessLifecycleObserver {
+
+    override suspend fun injectCookie(cookieValue: String?) = withContext(dispatcherProvider.io()) {
+        if (privacyProFeature.get().serpPromoCookie().isEnabled()) {
+            synchronized(cookieManager) {
+                kotlin.runCatching {
+                    val cookieString = "$SERP_PPRO_PROMO_COOKIE_NAME=${cookieValue.orEmpty()};HttpOnly;Path=/;"
+                    cookieManager.setCookie(HTTPS_WWW_SUBSCRIPTION_DDG_COM, cookieString)
+                }
+            }
+            return@withContext
+        }
+    }
+
+    override fun onStart(owner: LifecycleOwner) {
+        owner.lifecycleScope.launch(dispatcherProvider.io()) {
+            if (privacyProFeature.get().serpPromoCookie().isEnabled()) {
+                kotlin.runCatching {
+                    val cookies = cookieManager.getCookie(HTTPS_WWW_SUBSCRIPTION_DDG_COM) ?: ""
+                    val pproCookies = cookies.split(";").filter { it.contains(SERP_PPRO_PROMO_COOKIE_NAME) }
+                    if (pproCookies.isEmpty()) {
+                        injectCookie("")
+                    }
+                }
+            }
+        }
+    }
+}
+
+// This class is basically a convenience wrapper for easier testing
+interface CookieManagerWrapper {
+    /**
+     * @return the cookie stored for the given [url] if any, null otherwise
+     */
+    fun getCookie(url: String): String?
+
+    fun setCookie(url: String, cookieString: String)
+}
+
+@Retention(AnnotationRetention.BINARY)
+@Qualifier
+private annotation class InternalApi
+
+@Module
+@ContributesTo(AppScope::class)
+object CookieManagerWrapperModule {
+    @Provides
+    @InternalApi
+    fun providesCookieManagerWrapper(): CookieManagerWrapper {
+        return CookieManagerWrapperImpl()
+    }
+}
+private class CookieManagerWrapperImpl constructor() : CookieManagerWrapper {
```

Comment:
We already have a `CookieManagerWrapperImpl` inside `network-protection-impl`, rather than creating a new one perhaps we should move it to the cookies module?

Reply:
Yeah, after this I am gonna propose a `DuckDuckGoCookieManager` API and move all the related code to that